### ISSUE TITLE
Proposal: Improving formspec buttons

### ISSUE THREAD

**Issue by Df458** [minetest/minetest#9044]
Following some discussion in #8916, I've decided that my push to get the changes within merged was premature and for that I must apologize. To take responsibility for this mistake, I propose the following improvements, which will be split into 5 stages (each with its own PR). If any coredevs are willing to allow them, I will implement them myself with the goal of merging before 5.2.0.
**EDIT**: No dice, but we're making good progress nonetheless!

*Some of these changes will impact existing behavior*, so please read carefully. However, I believe that all changes here are for the better.

- [x] **Refactor `GUIButton` and make `image_button`/`item_image_button` consistent with `button`**
      Right now, `image_button` and `item_image_button` handle style elements separately from `button` and `item_image_button` is implemented rather strangely. The goal of this stage is threefold: Handle `image_button`/`item_image_button` as subclasses of GUIButton, centralize style parsing for all three elements, and add support for `bgcolor`/`bgimg` properties in all button variants. In other words, handle the images in `image_button`/`item_image_button` as *content* rather than as a *background*, allowing `style_type` prepends to skin them with a background texture without removing their content. This also makes `image_button` clearly useful again, right now styles have made it more-or-less redundant.
- [x] **Add state-selection to `style` and `style_type`, and expose button content to `style`**
      The new default behavior would be generally good, but some modders may want to avoid/override it. to make this easy, I propose adding conditional styles based on the state of the element (hovered and pressed initially, with more to come later). At the same time, new style properties like `text`, and `image` could be added to the respective button elements. I have a couple ideas for state selection, I will present them at greater length when the time comes.
- [x] **Support 9-slice button backgrounds**
      Buttons can be sized arbitrarily, in order to ensure that they look good I think support for 9-slice textures is a no-brainer. this sort of change only makes sense when there's a distinction between 'foreground' and 'background' images, which is why I've added it at the end.
- [x] **Add the concept of custom offsets and padding to buttons** 
      Padding would allow buttons to display their content within any custom border, allowing `image_button` and `item_image_button` elements to really look nice with the new 9-slice backgrounds. This is considered an issue by modders (in fact, it was reported as a bug in #9615), and it's a sensible feature. We should add a default padding value based on the border width, and also allow modders to modify this value as needed.
      **Custom offsets** are a somewhat related, but different concept. Right now, button content is always offest by a static amount. With the advent of custom button backgrounds, we should allow the modder to alter this value. It would be useful for buttons with more 'depth' to them, as well as button designs where the content should not move.
- [ ] ~~**Improve default behavior of borderless buttons**
      Right now, buttons with a border automatically get their border lightened/darkened when no override is specified. This change would update the default behavior to do the same to borderless buttons. The new default would look like this:
        - Has border, no state overrides: auto lighten/darken 'border' background
        - No border, bg image, no state overrides: auto lighten/darken bg image
        - No border, no bg image, text/image content: auto lighten/darken content
      This would change default behavior, so it may be a little contentious. However, I believe it would make formspec UIs clearer and more consistent.~~
- [x] **Make `bgcolor_*` and `bgimg_*` combine properly** (#9818)
      Right now, the `bgcolor_*` properties only work on the default 'border' supplied by the button element. They should work everywhere, for two reasons. First, this is a more consistent and less confusing behavior. Second, this enables game creators (or even us, if we wanted!) to do something like `style_type[button,image_button,item_image_button;bgimg=custom_bg.png]` in a prepend and know that any instance of buttons using `bgcolor` will work as intended. It also means people don't have to sprinkle as many image modifiers into their formspecs, which I consider a bonus.
- [ ] **Add `active` and `disabled` states, and expose `disabled` state in formspec elements**
      Support for these states is one thing that was intentionally missing from my initial PR. This would add them, as well as a way to create disabled formspec elements (useful for things like graying-out the play button when no world is selected, for instance)

If all of these changes are implemented, I think formspec buttons will be able to achieve much better UX in general. Once these changes are implemented, we can also perform similar revamps of all other formspec elements in turn. I'm sure some people will argue for more comprehensive support right away, but since `button` is the only element with visual feedback we can temporarily hold off on doing so in order to accelerate the review/merge process and ensure that these changes make it in.

What are your thoughts? I will keep this list up to date as development continues.

**Comment by SmallJoker** [comment]
> I've decided that my push to get the changes within merged was premature and for that I must apologize.

You gave your hand and they took the entire arm. Nothing to apologize, although I'm very happy to see these plans for better buttons.

**Comment by Wuzzy2** [comment]
I would add to the list that all button types be colorizable. Currently, only the “normal” buttons are colorizable. The other button types like image button cannot be colorized yet.

It seems number 1 kinda implies that, just wanted to write it down more explicitly.

**Comment by Df458** [comment]
> It seems number 1 kinda implies that, just wanted to write it down more explicitly.

Your assumption is correct. I've adjusted the text for clarity.

**Comment by rubenwardy** [comment]
Suggestion: remove support for borderless, instead use a default 9-slice bgimg

**Comment by Df458** [comment]
> Suggestion: remove support for borderless, instead use a default 9-slice bgimg. border=false would clear the bgimg

Could you expand on that? If you mean introducing a bgimg to replace the current 'default look', then we need to make bgcolor affect bgimg because it currently doesn't. This missing interaction was pointed out by @kilbith a while back, I was planning to look into it when addressing the "Improve default behavior" task.

**Comment by rubenwardy** [comment]
I was thinking it could be beneficial in terms of removing engine hardcoded ways of styling, but it would probably end up with more weirdness in terms of handling `border` etc

**Comment by v-rob** [comment]
Just wondering, but why is the `enabled` state planned instead of the `disabled` state?  It seems that we would want to change properties specifically for when it's disabled, not have to add the `enabled` state to everything.

With `enabled`:
```
-- Normal buttons:
style_type[button:enabled;...]
style_type[button:hovered+enabled;...]
style_type[button:pressed+enabled;...]
style_type[button:hovered+pressed+enabled;...]
-- Disabled buttons:
style_type[button;...]
```

With `disabled`:
```
-- Normal buttons:
style_type[button;...]
style_type[button:hovered;...]
style_type[button:pressed;...]
style_type[button:hovered+pressed;...]
-- Disabled buttons:
style_type[button:disabled;...]
```

`disabled` is a lot easier IMO because it's a less common use-case.

**Comment by Df458** [comment]
Because the design of state selection was different when I wrote the proposal. You're completely right about `disabled`.

**Comment by v-rob** [comment]
I was thinking: Since we're going to add `:checked` at some point, why don't we add something for buttons to minimize the code we have to implement for checkboxes add add togglebutton support?

Irrlicht already supports this with `setIsPushButton` and `setPressed`, so it would just be a matter of adding `:checked` styling to that.  To make a button a togglebutton, perhaps we could do, say, `set_togglebutton[<button name>;<is toggled>]`

Just wanted to throw that out there.

**Comment by Df458** [comment]
As I've continued working on making button formspec APIs more flexible and less hard-coded, I've become less and less sure of the "Improve default behavior of borderless buttons" task. I think I'm going to do away with it in favor of just making `bgcolor` and `bgimg` play nice, as this is the most important (and least questionable) aspect of the task.

Being able to recolor background images, resize content, change offsets, etc. makes working out if feedback is present or not difficult, and gives the user a lot of freedom to make good feedback without necessarily redrawing background assets, so I don't think trying to work in special feedback rules outside of the most basic default case is worthwhile.

**Comment by Df458** [comment]
As for your suggestions, @v-rob, these are both good features that I suppotrt but I don't plan to implement them in the short-term. My reasoning is quite simple: Users can _already_ make togglebuttons and radiobuttons through manual formspec work, and I'm primarily focusing on styling in this issue rather than new elements.

That doesn't mean that they're not important or useful for users, but my primary goal right now is to expand styling and help build a flexible API that allows users to create beautiful GUIs with good UX for their mods and games. As that becomes more and more possible, I'll be gradually shifting my focus to new elements, layouts, bindings/reactivity, etc. But not yet!

I would be very happy to see you or anyone else work on such a feature in parallel though :)

**Comment by v-rob** [comment]
Well, since Df458 has left (see https://github.com/minetest/minetest/pull/10222#issuecomment-670932770 if you didn't know), I suppose I shall take up the reigns for this.  #9818 seems to be already fine and just needs one more approval, so I probably don't need to intervene there.  So I'll just implement the last thing on the list, namely `active`/`disabled` states.

I feel that since multiple people have left Minetest recently with no prior warning, I'd better clarify that unless something really drastic happens, I don't see myself leaving Minetest anytime soon (especially with 11 open formspec PRs and soon to be 12!).  I really like contributing to Minetest, and, if nothing else, it's a lot of fun and a great learning experience.

**Comment by Df458** [comment]
Hi, quick note--I'm planning to delete my github account in the near future. The UI indicates that my issues will be deleted when I do, so if you want to keep this open please migrate any relevant info to a new issue and close this one.

I will leave this account up until **May 1st 2024**, so you have time to make this change.
